Python/immutability.c

```diff
@@ -156,281 +239,877 @@ static bool is_c_wrapper(PyObject* obj){
+        _Py_CLEAR_IMMUTABLE(obj);
+        // Return to the GC.
+        return_to_gc(obj);
+        return;
     }
-    struct _gc_runtime_state* gc_state = get_gc_state();
-    // Use `old[0]` here, we are setting the visited space to 0 in add_visited_set().
-    gc_list_merge(&(state->visited), &(gc_state->old[0].head));
+    debug_obj("Unfreezing %s @ %p\n", obj);
+    // Note: We don't need the details of the SCC for a simple unfreeze.
+    struct SCCDetails scc_details;
+    scc_reset_root_refcount(obj);
+    scc_add_internal_refcounts(obj, &scc_details);
+    scc_make_mutable(obj);
+    scc_return_to_gc(obj, true);
+}
 
 
-    PyGC_Head *next;
-    for (gc = _PyGCHead_NEXT(&(state->visited_untracked)); gc != &(state->visited_untracked); gc = next) {
-        next = _PyGCHead_NEXT(gc);
-        _Py_CLEAR_IMMUTABLE(_Py_FROM_GC(gc));
-        // Object was not tracked in the GC, so we don't need to merge it back.
-        _PyGCHead_SET_PREV(gc, NULL);
-        _PyGCHead_SET_NEXT(gc, NULL);
+static void unfreeze_and_finalize_scc(PyObject* obj)
+{
+    struct SCCDetails scc_details;
+    debug_obj("Unfreezing and finalizing SCC starting at %s @ %p rc = %zd\n", obj, Py_REFCNT(obj));
+
+    scc_set_refcounts_to_one(obj);
+    scc_add_internal_refcounts(obj, &scc_details);
+
+    // These are cases that we don't handle.  Return the state as mutable to the
+    // cycle detector to handle.
+    // TODO(Immutable): Lift the weak references to be handled here.
+    if (scc_details.has_weakreferences > 0 || scc_details.has_legacy_finalizers > 0) {
+        debug("There are weak references or legacy finalizers in the SCC.  Let cycle detector handle this case.\n");
+        debug("Weak references: %d, Legacy finalizers: %d\n", scc_details.has_weakreferences, scc_details.has_legacy_finalizers);
+        scc_make_mutable(obj);
+        scc_return_to_gc(obj, true);
+        return;
     }
-#endif
 
-    if (state->visited_list == NULL) {
-        return; // Nothing to do
+    // But leave cyclic list in place for the SCC.
+    scc_make_mutable(obj);
+
+    PyObject* n = obj;
+    if (scc_details.has_finalizers) {
+        // Call the finalizers for all objects in the SCC.
+        do {
+            PyObject* c = n;
+            n = scc_next(c);
+            if (_PyGC_FINALIZED(c))
+                continue;
+            destructor finalize = Py_TYPE(c)->tp_finalize;
+            if (finalize == NULL)
+                continue;
+            // Call the finalizer for the object.
+            finalize(c);
+            // Mark so we don't finalize it again.
+            _PyGC_SET_FINALIZED(c);
+        } while (n != obj);
     }
 
-    while (PyList_Size(state->visited_list) > 0) {
-        // Pop doesn't return a newref, but we know the object is still live
-        // as we didn't change anything.
-        PyObject* item = pop(state->visited_list);
-        _Py_CLEAR_IMMUTABLE(item);
+    // tp_clear all elements in the cycle.
```

Comment:
A comment to explain why this is safe would be good :)

```suggestion
    // tp_clear all elements in the cycle.
    // The elements will stay allocated since until `scc_return_to_gc` which will
    // decrement the RC by one
```